aws_lambda_powertools/utilities/idempotency/persistence/redis.py

```diff
@@ -0,0 +1,440 @@
+            # The idempotency key exist:
+            #    - previous invocation with the same key and not expired(active idempotency)
+            #    - previous invocation timed out (Orphan Record)
+            #    - previous invocation record expired but not deleted by Redis (Orphan Record)
+
+            idempotency_record = self._get_record(item["name"])
+
+            # status is completed and expiry_attr timestamp still larger than current timestamp
+            # found a valid completed record
+            if idempotency_record.status == STATUS_CONSTANTS["COMPLETED"] and not idempotency_record.is_expired:
+                raise IdempotencyItemAlreadyExistsError
+
+            # in_progress_expiry_attr exist means status is in_progress, and still larger than current timestamp,
+            # found a vaild in_progress record
+            if (
+                idempotency_record.status == STATUS_CONSTANTS["INPROGRESS"]
+                and idempotency_record.in_progress_expiry_timestamp
+                and idempotency_record.in_progress_expiry_timestamp > int(now.timestamp() * 1000)
+            ):
+                raise IdempotencyItemAlreadyExistsError
+
+            # If the code reaches here means we found an Orphan record.
+            raise IdempotencyOrphanRecordError
+
+        except IdempotencyOrphanRecordError:
+            # deal with orphan record here
+            # aquire a lock for default 10 seconds
+            with self._acquire_lock(name=item["name"]):
+                self.client.set(name=item["name"], value=encoded_item, ex=ttl)
+
+            # lock was not removed here intentionally. Prevent another orphan operation in race condition.
+        except (redis.exceptions.RedisError, redis.exceptions.RedisClusterException) as e:
+            raise e
+        except Exception as e:
+            logger.debug(f"encountered non-redis exception:{e}")
+            raise e
+
+    @contextmanager
+    def _acquire_lock(self, name: str):
+        """
+        aquire a lock for default 10 seconds
+        """
+        try:
+            acquired = self.client.set(name=f"{name}:lock", value="True", ex=self._orphan_lock_timeout, nx=True)
+            logger.debug("acquiring lock to overwrite orphan record")
+            if acquired:
+                logger.debug("lock acquired")
+                yield
+            else:
+                # lock failed to aquire, means encountered a race condition. Just return
+                logger.debug("lock failed to acquire, raise to retry")
+                raise IdempotencyItemAlreadyExistsError
+
+        finally:
+            ...
+
+    def _put_record(self, data_record: DataRecord) -> None:
+        if data_record.status == STATUS_CONSTANTS["INPROGRESS"]:
+            self._put_in_progress_record(data_record=data_record)
+        else:
+            # current this function only support set in_progress. set complete should use update_record
+            raise NotImplementedError
```

Comment:
This looks like a leaky abstraction. What about calling `update_record` instead of raising?